plugin/pkg/admission/podresourceallocation/admission.go

```diff
@@ -0,0 +1,199 @@
+}
+
+var _ admission.ValidationInterface = &Plugin{}
+
+// NewPodResourceAllocation creates a new instance of the PodResourceAllocation admission controller
+func NewPodResourceAllocation() *Plugin {
+	return &Plugin{
+		Handler:        admission.NewHandler(admission.Create, admission.Update),
+		nodeidentifier: nodeidentifier.NewDefaultNodeIdentifier(),
+	}
+}
+
+// Admit may mutate the pod's ResourcesAllocated or ResizePolicy fields for compatibility with older client versions
+func (p *Plugin) Admit(ctx context.Context, attributes admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		return nil
+	}
+	// Ignore all calls to subresources or resources other than pods.
+	if len(attributes.GetSubresource()) != 0 || attributes.GetResource().GroupResource() != api.Resource("pods") {
+		return nil
+	}
+
+	pod, ok := attributes.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	op := attributes.GetOperation()
+	// mutation to set defaults is done in admission plugin rather than in defaults
+	// because doing it here allows us to support create/update from older client versions
+	switch op {
+	case admission.Create:
+		podQoS := qos.GetPodQOS(pod)
+		for i := range pod.Spec.Containers {
+			// set container resize policy to nil for best-effort class
+			if podQoS == api.PodQOSBestEffort {
+				pod.Spec.Containers[i].ResizePolicy = nil
+				continue
+			}
+			// set resources allocated equal to requests, if not specified
+			if pod.Spec.Containers[i].Resources.Requests != nil {
+				if pod.Spec.Containers[i].ResourcesAllocated == nil {
+					pod.Spec.Containers[i].ResourcesAllocated = make(api.ResourceList)
+					for key, value := range pod.Spec.Containers[i].Resources.Requests {
+						pod.Spec.Containers[i].ResourcesAllocated[key] = value.DeepCopy()
+					}
+				}
+			}
+			// set resize policy to defaults, if not specified
+			resources := make(map[api.ResourceName]bool)
+			for _, p := range pod.Spec.Containers[i].ResizePolicy {
+				resources[p.ResourceName] = true
+			}
+			if _, found := resources[api.ResourceCPU]; !found {
+				pod.Spec.Containers[i].ResizePolicy = append(pod.Spec.Containers[i].ResizePolicy,
+					api.ResizePolicy{
+						ResourceName: api.ResourceCPU,
+						Policy:       api.NoRestart,
+					})
+			}
+			if _, found := resources[api.ResourceMemory]; !found {
+				pod.Spec.Containers[i].ResizePolicy = append(pod.Spec.Containers[i].ResizePolicy,
+					api.ResizePolicy{
+						ResourceName: api.ResourceMemory,
+						Policy:       api.NoRestart,
+					})
+			}
+		}
+
+	case admission.Update:
+		oldPod, ok := attributes.GetOldObject().(*api.Pod)
+		if !ok {
+			return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+		}
+		if len(pod.Spec.Containers) != len(oldPod.Spec.Containers) {
```

Comment:
slight preference for only doing the resource preservation when the length is the same, and letting a length change continue on and get the normal validation error about not changing pod content

```go
if len(pod.Spec.Containers) == len(oldPod.Spec.Containers) {
  ... preserve/populate values
}
```